crates/oxc_linter/src/rules/eslint/no_unneeded_ternary.rs

```diff
@@ -0,0 +1,280 @@
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+fn no_unneeded_ternary_diagnostic(span: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn("Unnecessary use of boolean literals in conditional expression")
+        .with_help("Remove this ternary operator")
+        .with_label(span)
+}
+
+fn no_unneeded_ternary_conditional_expression_diagnostic(span: Span) -> OxcDiagnostic {
+    // See <https://oxc.rs/docs/contribute/linter/adding-rules.html#diagnostics> for details
+    OxcDiagnostic::warn("Unnecessary use of conditional expression for default assignment")
+        .with_help("Remove this ternary operator and use the variable directly")
+        .with_label(span)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoUnneededTernary {
+    default_assignment: bool,
+}
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow ternary operators when simpler alternatives exist
+    ///
+    /// ### Why is this bad?
+    ///
+    /// It’s a common mistake in JavaScript to use a conditional expression to select between two
+    /// Boolean values instead of using ! to convert the test to a Boolean.
+    ///
+    /// Another common mistake is using a single variable as both the conditional test and the
+    /// consequent. In such cases, the logical OR can be used to provide the same functionality.
+    ///
+    /// ### Examples
+    ///
+    /// Examples of **incorrect** code for this rule:
+    /// ```js
+    /// const isYes = answer === 1 ? true : false;
+    /// const isNo = answer === 1 ? false : true;
+    ///
+    /// foo(bar ? bar : 1);
+    /// ```
+    ///
+    /// Examples of **correct** code for this rule:
+    /// ```js
+    /// const isYes = answer === 1;
+    /// const isNo = answer !== 1;
+    ///
+    /// foo(bar || 1);
+    /// ```
+    NoUnneededTernary,
+    eslint,
+    suspicious,
+    pending
+);
+
+impl Rule for NoUnneededTernary {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let default_assignment = value
+            .get(0)
+            .and_then(|v| v.get("defaultAssignment"))
+            .and_then(serde_json::Value::as_bool)
+            .unwrap_or(true);
+
+        Self { default_assignment }
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::ConditionalExpression(expr) = node.kind() else {
+            return;
+        };
+        if matches!(expr.consequent, Expression::BooleanLiteral(_))
+            && matches!(expr.alternate, Expression::BooleanLiteral(_))
+        {
+            ctx.diagnostic(no_unneeded_ternary_diagnostic(expr.span));
+        }
+
+        if let (Some(ze_test), Some(ze_cons)) =
+            (supply_identifier(&expr.test), supply_identifier(&expr.consequent))
+        {
+            if !self.default_assignment && ze_test.name == ze_cons.name {
+                ctx.diagnostic(no_unneeded_ternary_conditional_expression_diagnostic(expr.span));
+            }
+        }
+    }
```

Comment:
a few suggestions:

1. since these two cases are mutually exclusive, this second `if` block can be moved into an `else` case so that we don't do unnecessary work
2. the `supply_identifier` can be implemented using built-in methods: `get_inner_expression()` will remove parentheses and such, among other "transparent" expressions; `get_identifier_reference()` will return the ident, if one exists.

```suggestion
        } else if let (Some(test), Some(cons)) = (
            (&expr.test.get_inner_expression().get_identifier_reference()),
            (&expr.consequent.get_inner_expression().get_identifier_reference()),
        ) {
            if !self.default_assignment && test.name == cons.name {
                ctx.diagnostic(no_unneeded_ternary_conditional_expression_diagnostic(expr.span));
            }
        }
```

Reply:
thank you for your comment